cores/arduino/stm32/hw_config.c

```diff
@@ -59,6 +60,9 @@ void hw_config_init(void)
   /* Initialize the HAL */
   HAL_Init();
 
+  /* Check if a jump to system memory boot requested */
+  jumpToBootloader();
```

Comment:
Isn't this a bit late to jump into the bootloader? AFAICS, `premain` (that calls `hw_config_init` ultimately) and HAL_Init already initialize some hardware (enable caches, enable the systick timer) and does `dwt_init()` enable the watchdog timer? The bootloader is expecting a completely uninitialized environment, so maybe `jumpToBootloader()` should just be called at the start of `premain()` instead?

Reply:
Currently not as I use the `USBD_reenumerate` which require basic init but this should be better.

Reply:
Ah, right, as USBD_reenumerate might use GPIO functions of course. Might be good to have a lighter version of that, then. Tricky part is that `USBD_reenumerate` might be overridden by variants, which would then also need to provide the lighter version...

Reply:
Right that will be fine.

Reply:
For inspiration, the STM32L4 core inserts an *extra* reset handler in front of the normal one, by modifying the linker script: https://github.com/GrumpyOldPizza/arduino-STM32L4/blob/913a9115b9e2654c563e17b0aae552c25bb7713f/variants/STM32L432KC-Ladybug/linker_scripts/STM32L432KC_FLASH.ld#L72-L79

This was all a bit confusing (there was a reset handler defined in C in boot_stm32l432.c, and one defined in assembler) until I found the linker script that puts the C version first, then the the assembler version at offset 0x800 (2048), and found that the C version then jumps to the assembler version at https://github.com/GrumpyOldPizza/arduino-STM32L4/blob/913a9115b9e2654c563e17b0aae552c25bb7713f/system/STM32L4xx/Source/boot_stm32l4xx.c#L177-L183

To be honest, this all seems rather complicated and we probably should not use this as-is, but maybe there is some useful inspiration in there.

Reply:
The code works around quite a few HW issues for the various different L4 incarnations. As well as a few GCC bugs (hence the assembly code).

The double boot code is to have a bootloader that can reflash the main arduino sketch from an image downloaded into normal flash memory. Not really needed for your purposes I suppose.